src/Shell/Task/ModelTask.php

```diff
@@ -395,17 +427,69 @@ public function findBelongsToMany($model, array $associations)
                 $assocTable = substr($otherTableName, 0, $otherOffset);
             }
             if ($assocTable && in_array($assocTable, $tables)) {
-                $habtmName = $this->_camelize($assocTable);
-                $assoc = [
-                    'alias' => $habtmName,
-                    'foreignKey' => $foreignKey,
-                    'targetForeignKey' => $this->_modelKey($habtmName),
-                    'joinTable' => $otherTableName
-                ];
-                if ($assoc && $this->plugin) {
-                    $assoc['className'] = $this->plugin . '.' . $assoc['alias'];
+                $assocTableObject = TableRegistry::get($otherTableName);
+
+                $thisForeignKey = $primaryKey;
+                foreach ($thisForeignKey as $i => $fieldName) {
+                    if ($fieldName === 'id') {
+                        $thisForeignKey[$i] = Inflector::singularize($model->table()) . '_id';
+                    }
+                }
+                $otherForeignKeys = $assocTableObject->schema()->columns();
```

Comment:
I think this detection also does't work correctly. If the junction table contains some non-related foreign key, the association wouldn't appear in the generated table class.
```sql
CREATE TABLE users_sites(
  id INT NOT NULL AUTO_INCREMENT
, user_id INT NOT NULL 
, site_id INT NOT NULL
, role_id INT NOT NULL
, PRIMARY KEY (id)
, UNIQUE (user_id, site_id)
, FOREIGN KEY (user_id) REFERENCES users(id) -- For belongsToMany
, FOREIGN KEY (site_id) REFERENCES sites(id) -- For belongsToMany
, FOREIGN KEY (role_id) REFERENCES roles(id) -- For belongsTo
); 
```